### PR TITLE
Adds Subscriptions and VA API's Support

### PR DESCRIPTION
**Pull request by mayankamencherla**
- [x] Create Plan                
- [x] Fetch Plan                 
- [x] Fetch All plans            
- [x] Create subscription        
- [x] Fetch subscription         
- [x] Fetch all subscriptions    
- [x] Cancelling subscription    
- [x] Create addon               
- [x] Fetch addon                
- [x] Delete addon               
- [x] Create VA                  
- [x] Fetch VA                   
- [x] Edit VA                    
- [x] Fetch all VAs               
- [x] Close VA                    
- [x] Fetch VA payments           
- [x] Fetch payment bank transfer 
- [x] Create payment Transfer     
- [x] Fetch all payment transfers 
- [x] Create direct transfer    
- [x] Create transfer reversals 
- [x] Fetch transfer    
- [x] Edit transfer

### CODE REVIEW

**Comment by mayankamencherla** [comment]
@sunnyagg Can we just use edit VA for close? In API, we use the same route for both features.

File: src/Subscription.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        public Subscription Create(Dictionary<string, object> data)
+        {
+            string relativeUrl = GetEntityUrl();
+            List<Entity> entities = Request(relativeUrl, HttpMethod.Post, data);
+            return (Subscription)entities[0];
+        }
+
+        public Subscription Cancel()
+        {
+            string relativeUrl = GetEntityUrl() + "/" + this["id"] + "/cancel";
```
Comment (sunnyagg) [inline src/Subscription.cs]: Need to add constructor for this to create Subscription with given ID ?

File: src/VirtualAccount.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        public VirtualAccount Create(Dictionary<string, object> data)
+        {
+            string relativeUrl = "virtual_accounts";
+            List<Entity> entities = Request(relativeUrl, HttpMethod.Post, data);
+            return (VirtualAccount)entities[0];
+        }
+
+        public VirtualAccount Edit(Dictionary<string, object> data) 
+        {
+            string relativeUrl = string.Format("virtual_accounts/{0}", this["id"]);
```
Comment (sunnyagg) [inline src/VirtualAccount.cs]: same here as well

File: test/Helper.cs [hunk 3]
```diff
@@ -396,6 +397,161 @@ public static Transfer PatchTransferByIdTest()
...
+            Dictionary<string, object> data = new Dictionary<string, object>();
+
+            using (StreamReader r = new StreamReader("./test/data/addon-create.json"))
+            {
+                string json = r.ReadToEnd();
+                data = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
+            }
+
+            Subscription subscription = TestCreateSubscription();
+
```
Comment (sunnyagg) [inline test/Helper.cs]: Create a new subscription with the id of created subscription

File: test/Helper.cs [hunk 4]
```diff
@@ -396,6 +397,161 @@ public static Transfer PatchTransferByIdTest()
...
+            Addon addon = TestCreateAddon();
+
+            Addon addon2 = Helper.client.Addon.Fetch((string)addon["id"]);
+
+            Assert.AreEqual(addon["id"], addon2["id"]);
+
+            return addon2;
+        }
+
+        // TODO: Need to re-test this
```
Comment (sunnyagg) [inline test/Helper.cs]: remove ?

File: src/Addon.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            }
+
+            return addons;
+        }
+
+        public Addon Delete()
+        {
+            string relativeUrl = GetEntityUrl() + "/" + this["id"];
+            List<Entity> entities = Request(relativeUrl, HttpMethod.Delete, null);
+            return (Addon)entities[0];
```
Comment (sunnyagg) [inline src/Addon.cs]: this will throw NullRef.

File: test/Helper.cs [hunk 6]
```diff
@@ -396,6 +397,161 @@ public static Transfer PatchTransferByIdTest()
...
+        public static Addon TestDeleteAddon()
+        {
+            Addon addon = Helper.client.Addon.All()[0];
```
Comment (sunnyagg) [inline test/Helper.cs]: we should be able to delete from a addon id. need to add a constructor to create addon with given Id

File: test/Helper.cs [hunk 7]
```diff
@@ -396,6 +397,161 @@ public static Transfer PatchTransferByIdTest()
...
+
+            Dictionary<string, object> data = new Dictionary<string, object>();
+
+            using (StreamReader r = new StreamReader("./test/data/virtualaccount-edit.json"))
+            {
+                string json = r.ReadToEnd();
+                data = JsonConvert.DeserializeObject<Dictionary<string, object>>(json);
+            }
+
+            return va.Edit(data);
```
Comment (sunnyagg) [inline test/Helper.cs]: Edit by creating a new VA with id of newly created va

File: test/Helper.cs [hunk 8]
```diff
@@ -396,6 +397,161 @@ public static Transfer PatchTransferByIdTest()
...
+        }
+
+        public static VirtualAccount VirtualAccountCloseTest()
+        {
+            VirtualAccount va = VirtualAccountCreateTest();
+
+            return va.Close();
```
Comment (sunnyagg) [inline test/Helper.cs]: same here, close by creating a new va with id of newly created va

File: test/Helper.cs [hunk 9]
```diff
@@ -396,6 +397,161 @@ public static Transfer PatchTransferByIdTest()
...
+        public static List<VirtualAccount> FetchAllVirtualAccountTest()
+        {
+            return Helper.client.VirtualAccount.All();
+        }
+
+        public static List<Payment> FetchAllPaymentsFromVa()
+        {
+            VirtualAccount va = VirtualAccountCreateTest();
+
+            return va.Payments();
```
Comment (sunnyagg) [inline test/Helper.cs]: same here as well

File: test/Helper.cs [hunk 10]
```diff
@@ -396,6 +397,161 @@ public static Transfer PatchTransferByIdTest()
...
+        public static List<Subscription> TestFetchSubscriptions()
+        {
+            return Helper.client.Subscription.All();
+        }
+
+        public static Subscription TestCancelSubscription()
+        {
+            Subscription subscription = TestCreateSubscription();
+
+            return subscription.Cancel();
```
Comment (sunnyagg) [inline test/Helper.cs]: same here, cancel on a object created with the id of existing sub

File: src/Addon.cs (hunk 5 above)
Comment (mayankamencherla) [reply src/Addon.cs]: @sunnyagg 
```
public void Delete()
{
      string relativeUrl = GetEntityUrl() + "/" + this["id"];
      Request(relativeUrl, HttpMethod.Delete, null);
}
```

File: test/Helper.cs (hunk 3 above)
Comment (sunnyagg) [reply test/Helper.cs]: Still not fixed

**Pull request merged**